Src/FluentAssertions/Specialized/TaskCompletionSourceAssertions.cs

```diff
@@ -124,18 +124,18 @@ public TaskCompletionSourceAssertions(TaskCompletionSource<T> tcs, IClock clock)
             .BecauseOf(because, becauseArgs)
             .FailWith("Expected {context} to complete within {0}{reason}, but found <null>.", timeSpan);
 
-        if (!success)
+        if (success)
         {
-            return new AndWhichConstraint<TaskCompletionSourceAssertions<T>, T>(this, default(T));
+            bool completesWithinTimeout = await CompletesWithinTimeoutAsync(subject.Task, timeSpan);
+            Execute.Assertion
+                .ForCondition(completesWithinTimeout)
+                .BecauseOf(because, becauseArgs)
+                .FailWith("Expected {context:task} to complete within {0}{reason}.", timeSpan);
+            T result = subject.Task.IsCompleted ? subject.Task.Result : default;
+            return new AndWhichConstraint<TaskCompletionSourceAssertions<T>, T>(this, result);
         }
 
-        bool completesWithinTimeout = await CompletesWithinTimeoutAsync(subject.Task, timeSpan);
-        Execute.Assertion
-            .ForCondition(completesWithinTimeout)
-            .BecauseOf(because, becauseArgs)
-            .FailWith("Expected {context:task} to complete within {0}{reason}.", timeSpan);
-        T result = subject.Task.IsCompleted ? subject.Task.Result : default;
-        return new AndWhichConstraint<TaskCompletionSourceAssertions<T>, T>(this, result);
+        return new AndWhichConstraint<TaskCompletionSourceAssertions<T>, T>(this, default(T));
```

Comment:
Have you considered creating a small factory method on `AndWhichConstraint` and similar generic classes to avoid having to declare all the redundant/noisy type information explicitly?

Something like this would be significantly cleaner:
```cshsarp
return AndWhichConstraint.For(this, default(T));
```

Reply:
Oeh, I like that.